core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -498,6 +498,45 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
       requestLocal)
   }
 
+  /*
+    The state transition for Empty, CompleteAbort, CompleteCommit is complicated, so here is a table to make it clear.
+    Note:
+    PF = PRODUCER_FENCED
+    ITS = INVALID_TXN_STATE
+    NONE = No error and no epoch bump
+    EB = No error and epoch bump
+
+    Retry => producerEpoch = txnState.ProducerEpoch - 1
+    Current => producerEpoch = txnState.ProducerEpoch
+    ------------------------------------------------------
+    With transaction V1.
+    +----------------+-----------------+-----------------+
+    |                | Abort           | Commit          |
+    +----------------+-------+---------+-------+---------+
+    |                | Retry | Current | Retry | Current |
+    +----------------+-------+---------+-------+---------+
+    | Empty          | PF    | ITS     | PF    | ITS     |
+    +----------------+-------+---------+-------+---------+
+    | CompleteAbort  | PF    | NONE    | PF    | ITS     |
+    +----------------+-------+---------+-------+---------+
+    | CompleteCommit | PF    | ITS     | PF    | NONE    |
+    +----------------+-------+---------+-------+---------+
+
+    With transaction V2.
+    +----------------+-----------------+-----------------+
+    |                | Abort           | Commit          |
+    +----------------+-------+---------+-------+---------+
+    |                | Retry | Current | Retry | Current |
+    +----------------+-------+---------+-------+---------+
+    | Empty          | PF    | EB      | ITS   | ITS     |
```

Comment:
Yeah, I think retry here means epoch that is not the current epoch. (But it is a bit confusing to call it retry)

Reply:
Retry is a very special type of epoch not match which is current epoch - 1. I guess it can happen if the client does not receive the response.

> For commit, the "retry" case is "PF".

Why is that? I think the server can return the current epoch to the client right? The client may just resend a commit request.

Reply:
Discussed offline. Correct the (Empty, commit, Retry) to PF. Leave the (completeCommit, Retry, Commit) = ITS.